posts/2021/02/pytorch-tensoriterator-internals-update.md

```diff
@@ -0,0 +1,321 @@
+
+at::TensorIteratorConfig iter_config;
+auto iter = iter_config
+  .check_all_same_dtype(false)
+  .resize_outputs(false)
+  .declare_static_shape(self.sizes(), /*squash_dim=*/dim)
+  .add_output(result)
+  .add_input(self)
+  .build();
+
+// Size of dimension to calculate the cumulative sum across
+int64_t self_dim_size = at::native::ensure_nonempty_size(self, dim);
+
+// These strides indicate number of memory-contiguous elements, not bytes,
+// between each successive element in dimension `dim`.
+auto result_dim_stride = at::native::ensure_nonempty_stride(result, dim);
+auto self_dim_stride = at::native::ensure_nonempty_stride(self, dim);
+
+auto loop = [&](char** data, const int64_t* strides, int64_t n) {
+  // There are `n` individual vectors that span across dimension `dim`, so
+  // `n` is equal to the number of elements in `self` divided by the size of
+  // dimension `dim`.
+
+  // These are the byte strides that separate each vector that spans across
+  // dimension `dim`
+  auto* result_data_bytes = data[0];
+  const auto* self_data_bytes = data[1];
+
+  for (int64_t vector_idx = 0; vector_idx < n; ++vector_idx) {
+
+    // Calculate cumulative sum for each element of the vector
+    auto cumulative_sum = (at::acc_type<float, false>) 0;
+    for (int64_t elem_idx = 0; elem_idx < self_dim_size; ++elem_idx) {
+      const auto* self_data = reinterpret_cast<const float*>(self_data_bytes);
+      auto* result_data = reinterpret_cast<float*>(result_data_bytes);
+      cumulative_sum += self_data[elem_idx * self_dim_stride];
+      result_data[elem_idx * result_dim_stride] = (float)cumulative_sum;
+    }
+
+    // Go to the next vector
+    result_data_bytes += strides[0];
+    self_data_bytes += strides[1];
+  }
+};
+
+iter.for_each(loop);
+```
+
+### Helper functions
+
+There are many helper functions within PyTorch that can simplify the creation
+and execution of a `TensorIterator`. We cannot cover all of them in this blog
+post, so you would need to discover them on your own. However, let's discuss
+one of the most common ones:
+[`make_reduction`](https://github.com/pytorch/pytorch/blob/d9e6750759b78c68e7d98b80202c67bea7ba24ec/aten/src/ATen/native/ReduceOpsUtils.h#L196)
+
+`make_reduction` creates a `TensorIterator` specifically for a reduction
+operation with one input and one output. It handles all of the
+`TensorIteratorConfig` setup internally, so we don't need to write as much
+boiler-plate code.
+
+The following example uses `make_reduction` to create a `TensorIterator` which
+is used to calculate the sum reduction of a 2-D input across dimension 1. This
+is equivalent to `torch.sum(self, dim=1)` in Python. If we didn't use
+`make_reduction`, this code would be a bit more complex and more difficult to
+write.
+
+```cpp
+at::Tensor self = at::randn({10, 10});
+int64_t dim = 1;
+bool keepdim = false;
+
+// `make_reduction` will allocate result Tensor for us, so we
+// can leave it undefined
+at::Tensor result;
```

Comment:
What will happen if we feed a defined Tensor? Is it going to be resized or we can expect a segfault if it happens to be too small?

Reply:
I believe that the correct treatment of `out=` arguments could make for a fantastic addition to the post, as it is surprisingly tricky at time.

Reply:
`make_reduction` resizes the result tensor if it's defined and isn't the correct size. I can add a note about this--perhaps it would be instructive if I init `result` with an incorrect size in the example.

You're right that overlaps aren't being checked, so I'll also add a note about that as well.

Reply:
Ivan Yashchuk wrote a number of PRs making sure that the behaviour of the `out=` version was consistent across `torch.linalg` functions. It believe that it's done by using something like `at::native::resize_output(result, sizes);`. This is done by `TensorIterator` automatically unless one specifies `  .resize_outputs(false)`, as it is done in the other example. I do not know whether `make_reduction` handles this correctly, but by Nik's comment I assume it does not (?).

Reply:
While doing a review today I got into the `make_reduction` code and I found that it does not handle it well (it does not raise a warning if it resizes an `out` tensor that was fed with the wrong size. Hopefully that will be solved soon (perhaps in that same PR), so perhaps it's fine not to touch on that here, or simply mention that it is handled internally in `make_reduction`, and otherwise one would need to use `.resize_outputs` or `at::native::resize_output`.

Reply:
Ok, since `make_reduction`'s handling of out tensors may change soon, I agree that it's probably best not to attempt to cover it in detail here